### ISSUE TITLE
1. As a end user I need to have better performance

### ISSUE THREAD

**Issue by tristancabel** [geomstats/geomstats#920]
Performance et passage à l’échelle par rapport au nombre de données et la dimension :
 - Moyenne géométrique 100 matrices SPD. ok avec nilearn pour 10x10→pas 15x15. Données Réelles ≥ 60x60
https://travis-ci.org/github/geomstats/geomstats/builds/759593811

Une piste serait de regardé Numba (regarder https://ct.gitlabpages.inria.fr/gallery/kepler-py/kepler-py.html)

**Comment by nguigs** [comment]
> * Moyenne géométrique 100 matrices SPD. ok avec nilearn pour 10x10→pas 15x15.
see #913 for more details

**Comment by nguigs** [comment]
```python
import geomstats.backend as gs
from geomstats.geometry.hypersphere import HypersphereMetric
from timeit import timeit
x = gs.array([1., 0., 0., 0.])
v = gs.array([0., 2., 1.5, .4])
metric = HypersphereMetric(3)
print(timeit(lambda: metric.norm(v, x), number=1000)) #  0.11182429000018601
print(timeit(lambda: gs.linalg.norm(v), number=1000)) # 0.008933541999795125
```

**Comment by tristancabel** [comment]
- [ ] profiling
- [ ] numba

**Comment by ninamiolane** [comment]
- [ ] interaction with joblib and OpenBLAS

**Comment by ninamiolane** [comment]
Could using cache be an option?

https://gist.github.com/Susensio/61f4fee01150caaac1e10fc5f005eb75

**Comment by mvesin** [comment]
tests & basic benchmarks report for issue 913 :
https://github.com/mvesin/geomstats/blob/feature/test_profiling/profiling/out.issue913/README.md
* confidence in our profiling instrumentation with cProfile
* OMP_NUM_THREADS=1 good practice when using PyPi numpy package (why ? to be investigated)
* ... but does not fill the performance gap between geomstats and nilearn for this example
* better performance and scaling for this example with nilearn, tests suggest to investigate implementation and algorithmic aspects

**Comment by mvesin** [comment]
more tests & benchmarks for issue 913 : 
https://github.com/mvesin/geomstats/blob/feature/test_profiling/profiling/out.issue_913.commit_eac607b8/README.md

**Comment by xpennec** [comment]
For an example where the convergence will take much longer than with SPD matrices, we may go to the sphere. One ca take the following example: https://github.com/geomstats/geomstats/blob/master/examples/empirical_frechet_mean_uncertainty_sn.py

Calling empirical_frechet_var_bubble(n_samples, theta, dim, n_expectation) with theta = gs.pi / 2.0 - 0.000001 should need much more iterations and does not always converges. The closer theta is to pi/2, the more difficult it is. Other parameters can be taken as follows:  n_sample = 2 or 3, dim = 2 to 10 or larger (not tested),  n_expectation = 10 or more (this is just repeating the same experiment n_expectation times with different random samples).

**Comment by mvesin** [comment]
Issue #913, time per iteration benches : https://github.com/mvesin/geomstats/blob/feature/test_profiling/profiling/out.issue_913.04_commit_2595549b/README.md

**Comment by tristancabel** [comment]
next stuff to study is why *self.embedding_metric.norm* slower than *gs.linalg* .

**Comment by mvesin** [comment]
Issue #913 time per iteration benches on packaging of numpy (OpenBlas vs MKL) and threading (in blas) : https://github.com/mvesin/geomstats/tree/feature/test_profiling/profiling/out.issue_913.09_commit_9918a42f

**Comment by mvesin** [comment]
first bunch of tests for HypersphereMetric().norm performance issue : https://github.com/mvesin/geomstats/blob/feature/test_profiling/profiling/out.hsmetric-norm.01_commit_8d611afb/README.md
Suggest to continue investigating 3 axis with significant potential gains more or less cumulative :+1:
* geomstats vectorization decorators
* pure numpy vs autograd.numpy
* riemaniann_metric.py inner_product rewrite

**Comment by mvesin** [comment]
Updated #920 `HypersphereMetric.norm()` benches with current code from master branch : https://github.com/mvesin/geomstats/blob/feature/test_profiling/profiling/out.hsmetric-norm.03_commit_5bed571a/README.md
- norm() & metrics rewrite by @Nicolas does a great job for performance with numpy backend ! ~50-60% lower execution time, we are now only ~ x1.2-1.35 execution time of gs.linalg.norm()  with probably additional possible gains with a pure numpy backend (vs autograd.numpy current backend), but did not re-test for the current code version
- pytorch backend tests shows execution time for HypersphereMetric.norm() is ~ x12 that of gs.linalg.norm(). This is mainly due to geomstat's custom einsum in geomstats/_backend/pytorch/__init__.py as it represents >88% of execution time.
- tensorflow backend shows gs.linalg.norm() takes ~ x15-20 longer execution time than same function with other backends. So HypersphereMetric.norm() takes only ~ x1.3 longer, but as for pytorch, execution time is dominated (>92%) by geomstat's custom einsum in geomstats/_backend/tensorflow/__init__.py .
- so there is probably big potential gain from optimizing einsum for pytorch and tensorflow backends

**Comment by mvesin** [comment]
# Summary of performance tuning actions and findings

## test cases

(a) #913 : `FrechetMean` performance vs nilearn `_geometric_mean` performance
(b) #920 [code snippet](https://github.com/geomstats/geomstats/issues/920#issuecomment-792782040) : 10k iterations on `HypersphereMetric.norm()` vs `gs.linalg.norm()` performance

__WIP__
